services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,424 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.deviceregistry.mongodb.service;
+
+import java.net.HttpURLConnection;
+import java.util.Objects;
+import java.util.Optional;
+
+import javax.security.auth.x500.X500Principal;
+
+import org.eclipse.hono.client.ClientErrorException;
+import org.eclipse.hono.deviceregistry.mongodb.config.MongoDbBasedTenantsConfigProperties;
+import org.eclipse.hono.deviceregistry.mongodb.model.TenantDto;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbCallExecutor;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbDeviceRegistryUtils;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbDocumentBuilder;
+import org.eclipse.hono.deviceregistry.util.DeviceRegistryUtils;
+import org.eclipse.hono.deviceregistry.util.Versioned;
+import org.eclipse.hono.service.management.Id;
+import org.eclipse.hono.service.management.OperationResult;
+import org.eclipse.hono.service.management.Result;
+import org.eclipse.hono.service.management.tenant.Tenant;
+import org.eclipse.hono.service.management.tenant.TenantManagementService;
+import org.eclipse.hono.service.tenant.TenantService;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.AuthenticationConstants;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.eclipse.hono.util.RegistryManagementConstants;
+import org.eclipse.hono.util.TenantResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import io.opentracing.Span;
+import io.opentracing.noop.NoopSpan;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.FindOptions;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.UpdateOptions;
+
+/**
+ * An implementation of the tenant service and the tenant management service
+ * that uses a mongodb database to store tenants.
+ * <p>
+ * On startup this adapter tries to find the tenant collection, if not found, it gets created.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/tenant/">Tenant API</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/management/">Device Registry Management API</a>
+ */
+@Component
+@Qualifier("serviceImpl")
+public final class MongoDbBasedTenantService extends AbstractVerticle
+        implements TenantService, TenantManagementService {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDbBasedTenantService.class);
```

Comment:
As per the naming conventions, we could change this static final variable to upper case (`LOG`).

Reply:
done